Source/ACE.Database/WorldDatabase.cs

```diff
@@ -376,6 +377,14 @@ public int GetPointsOfInterestCacheCount()
             return cachedPointsOfInterest.Count(r => r.Value != null);
         }
 
+        /// <summary>
+        /// Returns the PointsOfInterest cache.
+        /// </summary>
+        public ConcurrentDictionary<string, PointsOfInterest> GetPointsOfInterestCache()
```

Comment:
We shouldn't expose the base cache object.

Instead, if you want the elements, you should create a copy. We want to make sure a consumer doesn't have direct access to our base cache objects. They could end up doing something we don't want.

Reply:
I debated this with myself, and another developer and we agreed that it should probably return the meat of it instead of only a certain property in case someone wants to do something more like show the coordinates in the list as well.  I started out having it return `IEnumerable<string>` and switched to `ConcurrentDictionary<string, PointsOfInterest>` after thinking about it further.  I understand the need to prevent data from being tainted but really IMHO none of the consumers should be doing that, and if they are it's a bug.  The alternative is to write accessors for every specific type of application within that layer which I'm sure you would agree is crap, or like you said copy the objects, which is almost as crappy.  If it were up to me I would not copy it and just remember not to try and do any write operations to it from other layers.  I'll make this change, just remember that I disagree with your philosophy.

Reply:
We don't need to deep copy the elements, we just don't want to return access to the base cache dictionary.

All of the various objects returned by the various world database caching mechanisms are the same, not clones.

Reply:
I think a simple line like this will suffice:

return new Dictionary<string, PointsOfInterest>(cachedPointsOfInterest);

Reply:
I didn't know that, should have asked.  Thanks, I'll handle it.

Reply:
The individual elements will be shared by anyone using the cache, and yes, vulnerable to mutation.

This is how all of our world db cache works.

If you want to mutate those objects, you should then clone them. This would be the responsibility of the consumer, not the cache (as it is currently designed).

For weenies, we clone them into biotas and then mutate the biotas.